[jit] submodule swapping via module interface

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28409 [jit] submodule swapping via module interface**

Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18284309](https://our.internmc.facebook.com/intern/diff/D18284309)

wanchaol committed:
[jit] submodule swapping via module interface

Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the attribute as module interface type
but still keep the value slot as EntityType::MODULE so that JitType,
CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the attribute as module interface type
but still keep the value slot as EntityType::MODULE so that JitType,
CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
rebase with master on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the attribute as module interface type
but still keep the value slot as EntityType::MODULE so that JitType,
CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the attribute as module interface type
but still keep the value slot as EntityType::MODULE so that JitType,
CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito commented on torch/jit/_recursive.py:
```diff
@@ -329,6 +340,10 @@ def init_fn(script_module):
                 cpp_module._register_parameter(name, orig_value, False)
             elif isinstance(orig_value, torch.jit.Attribute):
                 cpp_module._register_attribute(name, attr_type, orig_value.value)
+            elif isinstance(orig_value, Module):
```
This seems wrong. I feel like if this wasn't done before, we are compiling all submodules twice. @suo who understands this logic better.

wanchaol replied:
concreteType will record the "moduleInterface" modules as attributes instead of modules, so `concrete_type.get_module_names()` will not get that module and it only get compiled here.

zdevito commented on torch/jit/_recursive.py:
```diff
@@ -88,6 +90,17 @@ def infer_raw_concrete_type(nn_module):
         added_names.add(name)
 
     for name, item in nn_module._modules.items():
+        if name in class_annotations:
```
This logic needs to be carefully done to avoid being fragile. First, it should avoid using continue. Because of continue it forgets to add name to `added_names`, which is probably a bug. The handling of the submodule is also weird. There is the concrete submodule, which needs to be converted into script using the get_or_create_concrete_type, and there is the interface type which needs to be added as an attribute. This seems to skip handling the concrete_type for the actual module being stored inside this submodule. I think this leads to the problem later where all submodules get recursively scripted again.

zdevito commented on torch/jit/_recursive.py:
```diff
@@ -329,6 +340,10 @@ def init_fn(script_module):
                 cpp_module._register_parameter(name, orig_value, False)
             elif isinstance(orig_value, torch.jit.Attribute):
```
This copypasta makes this PR more confusing. There should only be one call to _register_attribute in this loop, and it should be clear that it is the fallthrough pathway. I.e. previously modules were going through _register_attribute and the original value is (presumably) the already scripted module.

zdevito reviewed (commented):
This is a good start but there are a number of issues related to the way we infer the correct types for submodules such that I don't think the conversion logic is correct. @suo can you take a look at this as well? It interacts with the concrete_type logic.

wanchaol commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1347,6 +1347,66 @@ matchTypeVariables(TypePtr formal, TypePtr actual, TypeEnv& type_env);
 // does not appear in `type_env`
 CAFFE2_API TypePtr tryEvalTypeVariables(TypePtr type, TypeEnv& type_env);
 
+struct InterfaceType;
```
the reason of it being moved is because classType::is_module() need to know interfaceType so that it knows the submember is a module or not. I can do forward declaration for InterfaceType but this might be more neat.

wanchaol commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1521,6 +1588,9 @@ struct CAFFE2_API ClassType : public NamedType {
     if (auto cls = getAttribute(slot)->cast<at::ClassType>()) {
       return cls->is_module();
     }
+    if (auto cls = getAttribute(slot)->cast<at::InterfaceType>()) {
```
Yeah I think moving the is_module() to Type may be a better solution and it avoid moving interfaceType around, i will do that.

wanchaol commented on torch/jit/__init__.py:
```diff
@@ -1716,8 +1732,18 @@ def __setattr__(self, attr, value):
+                # If the attr is an module interface type, it's guranteed that the module is
+                # not inlined in the graph and acting like the attribute way, in this case we
+                # allow swapping a new scripted module in.
+                if _is_module_interface(self._c, attr):
```
One thing that has to be happened in python is that we are associating every Script::Module with the actual python module in the `OrderedModuleDict`, the python module associated after module interface swap has to be updated as well.

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the attribute as module interface type
but still keep the value slot as EntityType::MODULE so that JitType,
CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the attribute as module interface type
but still keep the value slot as EntityType::MODULE so that JitType,
CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the attribute as module interface type
but still keep the value slot as EntityType::MODULE so that JitType,
CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the attribute as module interface type
but still keep the value slot as EntityType::MODULE so that JitType,
CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol commented on torch/csrc/jit/script/module.h:
```diff
@@ -184,6 +184,23 @@ struct TORCH_API Module {
+  void setattr(const std::string& name, IValue v) {
+    size_t slot = module_object()->type()->getAttributeSlot(name);
+    const TypePtr& expected = module_object()->type()->getAttribute(slot);
+    // TODO: give better error message when subtyping check fails, and python_str
```
@suo This is where I think we should give better error message when subtyping check fails, and `python_str` are just mangled names from the same nn.Module, we shouldn't expose the mangled names to the user.

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the attribute as module interface type
but still keep the value slot as EntityType::MODULE so that JitType,
CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1495,7 +1495,14 @@ struct CAFFE2_API ClassType : public NamedType {
+    TORCH_CHECK(
+      ty->isSubtypeOf(atype),
+      ty->python_str(),
+      " does not compatible with the type ",
```
is not

zdevito commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -139,8 +139,8 @@ bool ClassType::isSubtypeOfExt(const TypePtr rhs, std::ostream* why_not) const {
                    << "  (2) " << schema << "\n";
-          return false;
         }
+        return false;
```
this changed should be in the other PR

zdevito commented on torch/csrc/jit/script/concrete_module_type.h:
```diff
@@ -190,12 +193,18 @@ class VISIBILITY_HIDDEN ConcreteModuleType {
   struct ModuleInfo {
-    std::string name;
-    std::shared_ptr<ConcreteModuleType> meta;
+    ModuleInfo(std::string name, std::shared_ptr<ConcreteModuleType> meta)
```
In the case where this module is actually a module interface, what is meta set to? This seems redundant to me.

suo replied:
Yes, we need to be careful about the concepts here. It's not just redundant, the current code is not correct I think. For example, you could use a method that is defined on the `ConcreteModuleType` pyclass, but not in the interface.

This really an either/or relationship, either this is a concrete module instance with a `ConcreteModuleType`, or a module interface that is restricted to the methods defined in the interface. I think we need to split these concepts.

wanchaol replied:
I think `ConcreteModuleType` of the submodule here is just used to create the correct JIT type, in the case where the submodule is a module interface, the meta will still be set, but actually  it will be shadowed by the ModuleInterfaceType that we created when we creating the parent JIT class Type (ModuleInfo use type_ as source of truth for creating JIT type). I don't quite understand why we could use a method that is defined on `ConcreteModuleType` pyclass, everything after compilation should use cpp_module as source of truth (and it set the correct ModuleInterface jit type) and concrete_module should be invisible to user?

zdevito reviewed (approved):
This looks good to me, @suo can you look at the changes in _recursive.py to make sure the way the concrete type is being constructed is correct. In particular, I am not sure what add_module should be doing in the case where we are adding a module interface rather than a concrete module. Should ModuleInfo contain a ConcreteModuleType or just a raw type for the module? For interfaces there won't be a ConcreteModuleType

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -184,6 +184,23 @@ struct TORCH_API Module {
     return module_object()->setAttr(name, v);
   }
 
+  void setattr(const std::string& name, IValue v) {
```
The addition of this method is suspicious to me. What's the difference between `set_attribute` and `setattr`? It appears the only difference is that there is a subtyping check, which should exist on `set_attribute` as well (the current implementation of `set_attribute` is incorrect.)

wanchaol commented on torch/csrc/jit/script/module.h:
```diff
@@ -184,6 +184,23 @@ struct TORCH_API Module {
     return module_object()->setAttr(name, v);
   }
 
+  void setattr(const std::string& name, IValue v) {
```
This is to accommodate the changes that @zdevito is going to make for the [Module `setattr`](https://github.com/pytorch/pytorch/pull/28828/files#diff-cc763db9870942abe2fb372daa76489fR472), I could fix `set_attribute` and use the correct logic here, but that might cause merge conflicts when that one getting merged.

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18284309](https://our.internmc.facebook.com/intern/diff/D18284309)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18284309](https://our.internmc.facebook.com/intern/diff/D18284309)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18284309](https://our.internmc.facebook.com/intern/diff/D18284309)

[ghstack-poisoned]

suo reviewed (changes requested):
Nice, we're close!

wanchaol commented on torch/jit/_recursive.py:
```diff
@@ -340,20 +351,23 @@ def init_fn(script_module):
-            scripted = recursive_script(orig_value)
+            if isinstance(module_type, torch._C.InterfaceType):
+                # use the interface inference rule to compile the module
+                scripted = create_script_module(orig_value, infer_interface_methods_to_compile(module_type, orig_value))
```
`create_script_module()` is also used in `ScriptMeta.init_then_script`, which is causing recursive error if we would like to move it there. I will figure out something that is reasonable.

wanchaol commented on test/test_jit_py3.py:
```diff
@@ -374,6 +374,128 @@ def three(self, x):
                     # type: (Tensor) -> Tensor
                     return 3 * x
 
+    def test_module_interface_swap(self):
```
For more tests, like item 2 and 3 are already exist in the https://github.com/pytorch/pytorch/pull/28408, I will just move them together to the new test file under test/jit

wanchaol replied:
Actually for the move of the tests from https://github.com/pytorch/pytorch/pull/28408 to test/jit/test_module_interface, I will do it in a follow up PR as this is getting too big and hard to review together with the previous changes.

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18284309](https://our.internmc.facebook.com/intern/diff/D18284309)

[ghstack-poisoned]

suo reviewed (approved):
shipit!

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18284309](https://our.internmc.facebook.com/intern/diff/D18284309)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18284309](https://our.internmc.facebook.com/intern/diff/D18284309)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] submodule swapping via module interface"


Summary:
This PR enables submodule swapping via module interface. User could
declare a submodule as an module interface type in the ScriptModule,
during compilation we will record the module interface type in
ModuleInfo of ConcreteModuleType, the JIT type associated will have the
correct ModuleInterfaceType, and CppModule will get the correct module list

Given that we still keep the module interface type in the type system,
the graph is not inlined when we call Module::Attr and it will use
prim::CallMethod to call the method, this allow us to do module swapping
for the ScriptModule that also meet the same module interface type, and
    we only allow the module swapping through the module interface
    approach.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18284309](https://our.internmc.facebook.com/intern/diff/D18284309)

[ghstack-poisoned]
